src/util/finite_field.cpp

```diff
@@ -0,0 +1,130 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Alex Ozdemir
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * A finite-field element, implemented as a wrapper around Integer.
+ */
+
+#include "util/finite_field.h"
+
+#include "base/exception.h"
+
+namespace cvc5::internal {
+
+const Integer& FiniteField::getValue() const { return d_value; }
+
+const Integer& FiniteField::getFieldSize() const { return d_size; }
+
+Integer FiniteField::toInteger() const { return d_value; }
+
+Integer FiniteField::toSignedInteger() const
+{
+  Integer half_size = d_size.divByPow2(1) + 1;
+  return (d_value < half_size) ? d_value : d_value - d_size;
+}
+
+std::string FiniteField::toString(unsigned int base) const
+{
+  return toSignedInteger().toString();
+}
+
+size_t FiniteField::hash() const { return d_value.hash() + d_size.hash(); }
```

Comment:
To combine hashes, use the `PairHashFunction`. Using `+` has the issue that if value and size are swapped, you get the same hash.

Reply:
https://github.com/cvc5/cvc5/pull/9049